### ISSUE TITLE
Explanation on (default) behaviour of pruning

### ISSUE THREAD

**Issue by mcrn-rocinante** [optuna/optuna#1529]
Hi, first of all i'd like to thank anyone who participated in this project.

I'm currently designing and training a neural network on pytorch for an imbalanced classification problem and I'm trying to use an hyperparameter optimizer such as optuna to see if it would improve the efficiency of my classifier.
Being new to pytorch and optuna my code is thus mostly "inspired" from this post: https://github.com/optuna/optuna/blob/master/examples/pytorch_simple.py

Due to the nature of my problem my evaluation of my model is not accuracy but instead i focus on precision and accuracy on the positive class, hence i created a custom criterion to optimize.

I wanted to know how pruning worked to understand how the whole process works, however reading the documentation wasn't enough.

Knowing that i use a trial.report() and trial.should_prune() at every epoch, how does the pruning lagorithm works:
Does it compare the value of my criterion with values from previous trials at same epoch or something else?
And if it does that, what proves that a trial that "started badly" won't actually be a good one (after the final number of epochs chosen).

If anyone could enlighten me, it would be of great help.
Thanks for reading and have a great day.
## Note to the questioner

If you are more comfortable with [Stack Overflow], you may consider posting your question with the "optuna" tag there instead.
Alternatively, for issues that would benefit from more of an interactive session with the developers,
you may refer to the [optuna/optuna chat] on Gitter.

[Stack Overflow]: https://stackoverflow.com/questions/tagged/optuna
[optuna/optuna chat]: https://gitter.im/optuna/optuna

**Comment by deng-cy** [comment]
The default pruner is ASHA/SHA. It compares with the other trials at the same epoch. You are right that it's possible a trial that "started badly" will actually be a good one. Then we need to be careful choosing how "early" we could make decisions to discard a trial, which is controlled by the hyperparameter `s` in SHA.  A safe solution is to loop over all possible `s`. This idea is called Hyperband, which is also included in Optuna.

**Comment by HideakiImamura** [comment]
Thanks for the question @mcrn-rocinante and thanks for the quick answer @deng-cy! As @deng-cy said, it is possible a trial that started badly will actually be a good one. To control the "aggressiveness" of pruning, we can use the following parameter for each pruner in Optuna.

#### `optuna.pruners.{MedianPruner, PercentilePruner, ThresholdPruner}`
The argument of `n_warmup_steps` can control the aggressiveness of pruning. If we set the bigger `n_warmup_steps`, we prune less aggressively. Pruning is disabled until the trial exceeds the given number of `n_warmup_steps`.

#### `optuna.pruners.SuccessiveHalvingPruner`
The argument of `min_early_stopping_rate` can control the aggressiveness of pruning. If we set the bigger `min_early_stopping_rate`, we prune less aggressively. This parameter is introduced in the [original paper](http://arxiv.org/abs/1810.05934). It is difficult to explain this parameter without referring to the algorithm detail, so I omit it here. Note that in Optuna we implement the ["ASHA (Asynchronous Successive Halving)"](http://arxiv.org/abs/1810.05934) as the `SuccessiveHalvingPruner`, not ["SHA"](https://arxiv.org/abs/1502.07943). ASHA is more appropriate than SHA for the distributed optimization setting, which is the main target of Optuna.

#### `optuna.pruners.HyperbandPruner`
In the latest version of Optuna's `HyperbandPruner`, there is no parameter to control the aggressiveness of pruning. The `HyperbandPruner` can automatically control the aggressiveness. The new and stable `HyperbandPruner` will be included in the upcoming v2.0 release. Please look forward to it!

**Comment by deng-cy** [comment]
Sorry I made a mistake there. I marked in the original comment. 
I am a little curious about the sentence "The HyperbandPruner can automatically control the aggressiveness." From the paper of Hyperband, it simply loops over all s without really "automatic control". Do you mean you implemented a upgraded version of Hyperband?

**Comment by HideakiImamura** [comment]
@deng-cy Thanks for the reply! As you mentioned, Hyperband checks several `min_early_stopping_rate "s"` by looping them. Hyperband does not choose `s` adaptively, it just checks several `s`. In Optuna implementation, we also adopt the feature. However, the maximum number of looped `s` is automatically determined by other arguments of Hyperband (reduction factor `\eta` and minimum resource `r`). So I said that "The HyperbandPruner can automatically control the aggressiveness."

[FYI] We implemented an improved version of Hyperband with the following characteristics to enhance the performance in the distributed optimization.
- Randomly determine the assigned bracket for a given trial.
- Adopt `ASHA` for each bracket, not `SHA`.

Now, I'm planning to more improve Hyperband in PR #1271 by introducing the "information passing mechanism between different brackets".

**Comment by mcrn-rocinante** [comment]
Thanks a lot for all these answers, it's clearer for me now.
